File: frontend/src/data/procedures.js
// Procedure hierarchy — 4 levels:
// Level 1: Category   (Obstetrics / Gynaecology / ART)
// Level 2: Type       (Antenatal / Intrapartum / Postnatal)
// Level 3: Procedure  (Vaginal Birth / Caesarean Section / …)
// Level 4: Detail     (SVD / Vacuum / Uncomplicated / …) — null means no L4
//
// Structure: { Cat: { Type: { Procedure: [details] | null } } }

export const PROC = {
  Obstetrics: {
    Antenatal: {
      ECV: null,
      'Cervical Cerclage': null,
    },
    Intrapartum: {
      'Vaginal Birth': ['SVD', 'Vacuum', 'Forceps'],
      'Caesarean Section': [
        'Uncomplicated',
        'Placenta Praevia',
        'Classical',
        'Multiple Pregnancy',
        'Abnormal Presentation',
        'Fully Dilated',
        'BMI > 50',
        'Other',
      ],
    },
    Postnatal: {
      'Perineal Repair': ['Episiotomy / 2nd Degree Tear', '3rd / 4th Degree Tear'],
      'Management of PPH': ['General', 'Bakri Balloon', 'B Lynch Suture', 'Other'],
      MROP: null,
    },
  },
  Gynaecology: {
    Hysteroscopy: {
      Diagnostic: null,
      'Endometrial Polyp': {
        'D&C/Polyp Forceps': null,
        Myosure: null,
        Resectoscope: null,
      },
      Fibroids: {
        'D&C/Polyp Forceps': null,
        Myosure: null,
        Resectoscope: null,
      },
      'Abnormal Uterine Anatomy': {
        Septum: null,
        'Scar Niche': null,
      },
      'Endometrial Ablation': {
        Novasure: null,
        Cavaterm: null,
        Resectoscope: null,
      },
      Other: null,
    },
    Laparoscopy: {
      Diagnostic: null,
      'Tubal Surgery': {
        Salpingectomy: null,
        Salpingostomy: null,
        'Tubal Ligation': null,
      },
      'Ovarian Surgery': {
        Detorsion: null,
        'Cyst Aspiration': null,
        Cystectomy: null,
        Oophorectomy: null,
      },
      Endometriosis: null,
      Myomectomy: null,
      Hysterectomy: {
        Subtotal: null,
        Total: null,
        'Vaginal Assisted': null,
      },
      Other: null,
    },
    Laparotomy: {
      'Tubal Surgery': null,
      'Ovarian Surgery': null,
      Hysterectomy: null,
      Myomectomy: null,
    },
    'Cervical Surgery': {
      LLETZ: null,
      'Cone Biopsy': null,
    },
    Urogynaecology: {
      TVT: null,
      Other: null,
    },
    'Vaginal Surgery': {
      'Anterior Repair': null,
      'Posterior Repair': null,
      'Vaginal Hysterectomy': null,
      SSF: null,
      Other: null,
    },
    'Perineal Surgery': {
      "Bartholinś Cyst": null,
      'Vulval Biopsy or Excision': null,
      'I&D': null,
      Perineoplasty: null,
      Other: null,
    },
    Contraception: {
      'Complex IUD Removal': null,
      'Complex Implanon Removal': null,
      'Insertion of IUD': null,
      'Insertion of Implanon': null,
    },
    'Pregnancy Management': {
      'Suction D&C for Miscarriage': null,
      'Termination of Pregnancy': {
        Medical: null,
        Surgical: null,
      },
    },
    Cystoscopy: null,
    'Surgical Wound Debridement': null,
  },
  ART: {
    'Oocyte Collection': {
      'Transvaginal Oocyte Collection': null,
      'Transabdominal Oocyte Collection': null,
    },
    'Embryo Transfer': null,
    'Male Reproductive Surgery': {
      'Open Testicular Biopsy': null,
      PESA: null,
      TESA: null,
      'Microsurgical Sperm Retrieval': null,
    },
  },
}

export const catBadge = {
  Obstetrics: 'b-obs',
  Gynaecology: 'b-gyn',
  ART: 'b-art',
}

export const catColors = {
  Obstetrics: '#fbbf24',
  Gynaecology: '#a78bfa',
  ART: '#2dd4bf',
}

export function getProcTypes(category) {
  if (!category || !PROC[category]) return []
  return Object.keys(PROC[category])
}

export function getProcProcedures(category, type) {
  if (!category || !type) return []
  const t = PROC[category]?.[type]
  if (!t || t === null) return []
  return Object.keys(t)
}

export function getProcDetails(category, type, procedure) {
  if (!category || !type || !procedure) return []
  const p = PROC[category]?.[type]?.[procedure]
  if (!p) return []
  if (Array.isArray(p)) return p
  if (typeof p === 'object') return Object.keys(p)
  return []
}

export function isTerminalType(category, type) {
  if (!category || !type) return false
  return PROC[category]?.[type] === null
}

File: backend/procedures.py
"""
PROC – Full procedure hierarchy for the ProLog fertility & gynaecology logbook.

Structure:
  Category -> Type -> Procedure -> Detail (or None)

When the value at any level is None it means that level has no sub-levels.
When a list is present at the Detail level it contains the selectable options.
"""

from typing import Any

PROC: dict[str, Any] = {
    "Obstetrics": {
        "Antenatal": {
            "ECV": None,
            "Cervical Cerclage": None,
        },
        "Intrapartum": {
            "Vaginal Birth": {
                "SVD": None,
                "Vacuum": None,
                "Forceps": None,
            },
            "Caesarean Section": {
                "Uncomplicated": None,
                "Placenta Praevia": None,
                "Classical": None,
                "Multiple Pregnancy": None,
                "Abnormal Presentation": None,
                "Fully Dilated": None,
                "BMI > 50": None,
                "Other": None,
            },
        },
        "Postnatal": {
            "Perineal Repair": {
                "Episiotomy / 2nd Degree Tear": None,
                "3rd / 4th Degree Tear": None,
            },
            "Management of PPH": {
                "General": None,
                "Bakri Balloon": None,
                "B Lynch Suture": None,
                "Other": None,
            },
            "MROP": None,
        },
    },

    "Gynaecology": {
        "Hysteroscopy": {
            "Diagnostic": None,
            "Endometrial Polyp": {
                "D&C/Polyp Forceps": None,
                "Myosure": None,
                "Resectoscope": None,
            },
            "Fibroids": {
                "D&C/Polyp Forceps": None,
                "Myosure": None,
                "Resectoscope": None,
            },
            "Abnormal Uterine Anatomy": {
                "Septum": None,
                "Scar Niche": None,
            },
            "Endometrial Ablation": {
                "Novasure": None,
                "Cavaterm": None,
                "Resectoscope": None,
            },
            "Other": None,
        },
        "Laparoscopy": {
            "Diagnostic": None,
            "Tubal Surgery": {
                "Salpingectomy": None,
                "Salpingostomy": None,
                "Tubal Ligation": None,
            },
            "Ovarian Surgery": {
                "Detorsion": None,
                "Cyst Aspiration": None,
                "Cystectomy": None,
                "Oophorectomy": None,
            },
            "Endometriosis": None,
            "Myomectomy": None,
            "Hysterectomy": {
                "Subtotal": None,
                "Total": None,
                "Vaginal Assisted": None,
            },
            "Other": None,
        },
        "Laparotomy": {
            "Tubal Surgery": None,
            "Ovarian Surgery": None,
            "Hysterectomy": None,
            "Myomectomy": None,
        },
        "Cervical Surgery": {
            "LLETZ": None,
            "Cone Biopsy": None,
        },
        "Urogynaecology": {
            "TVT": None,
            "Other": None,
        },
        "Vaginal Surgery": {
            "Anterior Repair": None,
            "Posterior Repair": None,
            "Vaginal Hysterectomy": None,
            "SSF": None,
            "Other": None,
        },
        "Perineal Surgery": {
            "Bartholin\u2019s Cyst": None,
            "Vulval Biopsy or Excision": None,
            "I&D": None,
            "Perineoplasty": None,
            "Other": None,
        },
        "Contraception": {
            "Complex IUD Removal": None,
            "Complex Implanon Removal": None,
            "Insertion of IUD": None,
            "Insertion of Implanon": None,
        },
        "Pregnancy Management": {
            "Suction D&C for Miscarriage": None,
            "Termination of Pregnancy": {
                "Medical": None,
                "Surgical": None,
            },
        },
        "Cystoscopy": None,
        "Surgical Wound Debridement": None,
    },

    "ART": {
        "Oocyte Collection": {
            "Transvaginal Oocyte Collection": None,
            "Transabdominal Oocyte Collection": None,
        },
        "Embryo Transfer": None,
        "Male Reproductive Surgery": {
            "Open Testicular Biopsy": None,
            "PESA": None,
            "TESA": None,
            "Microsurgical Sperm Retrieval": None,
        },
    },
}


def get_categories() -> list[str]:
    return list(PROC.keys())


def get_types(category: str) -> list[str] | None:
    cat = PROC.get(category)
    if cat is None or not isinstance(cat, dict):
        return None
    return list(cat.keys())


def get_procedures(category: str, type_: str) -> list[str] | None:
    cat = PROC.get(category)
    if not isinstance(cat, dict):
        return None
    typ = cat.get(type_)
    if typ is None or not isinstance(typ, dict):
        return None
    return list(typ.keys())


def get_details(category: str, type_: str, procedure: str) -> list[str] | None:
    cat = PROC.get(category)
    if not isinstance(cat, dict):
        return None
    typ = cat.get(type_)
    if not isinstance(typ, dict):
        return None
    proc = typ.get(procedure)
    if proc is None or not isinstance(proc, dict):
        return None
    return list(proc.keys())
